### PR TITLE
content-conformance: ensure-html-comments-on-own-line

### PR DESCRIPTION
**Pull request by thiskevinwang**
---

## Description

This adds a `ensure-html-comments-on-own-line` ContentConformance rule to ensure sure that HTML comments are authored on their own line, without any indentation.

```markdown
❌ No nesting in bullets / other primitives
* <!-- comment -->

❌ No indentation
*  <!-- comment -->

❌ No accidental usage in 4-space codeblocks
    <!-- comment -->
```

## PR Checklist 🚀

- [ ] Conduct thorough self-review.
- [ ] Add or update tests as appropriate.
- [ ] Write a useful description (above) to give reviewers appropriate context.
- [ ] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203847022231358

### CODE REVIEW

File: packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        // is HTML comment
+        if (node.value.includes('<!--')) {
+          // // if child of another node
+          if (node.position.start.column != 1) {
+            context.report(
+              'Detected an HTML comment that is likely nested. You may be using this in a bullet point or other markdown primitive, which is not permitted. Please ensure HTML comments are on their own line, without any indentation.',
+              file,
+              node
+            )
+          }
```
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js]: @BRKalow Added additional logic + test case: https://github.com/hashicorp/web-platform-packages/pull/166/commits/93573d9b5c1a7c2c4a46bf0d704606d31f9581ae

File: packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  description: 'Ensures that HTML comments are on their own line.',
+  executor: {
+    async contentFile(file, context) {
+      file.visit('jsx', (node) => {
+        // is HTML comment or JSX tag with HTML comment
+        if (node.value.includes('<!--')) {
+          // if child of another node
+          if (node.position.start.column != 1) {
+            context.report(
+              'Detected an HTML comment that is likely nested. You may be using this in a bullet point or other markdown primitive, which is not permitted. Please ensure HTML comments are on their own line, without any indentation or nesting.',
```
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js:15]: The reason (which currently only applies to TFC/TFE content): We detect "content-exclusion comments", or pairs of unnest HTML comments, in order to conditionally render blocks of markdown. This breaks down when nesting/indentation is used.

Is this info worth adding to the rule-message?

File: packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      // check for accidental 4-space indentation (codeblock)
+      file.visit('code', (node) => {
+        if (node.value.includes('<!--')) {
+          context.report(
+            'Detected an HTML comment in a code block. This is not permitted. Please ensure HTML comments are on their own line, without any indentation or nesting.',
+            file,
+            node
+          )
+        }
+      })
```
Comment (BRKalow) [inline packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js:50]: I'm a little nervous about this one, what if someone has a legitimate HTML comment as part of a codeblock? It seems unlikely, but it's technically not invalid.

Comment (thiskevinwang) [reply packages/content-conformance/src/rules/ensure-html-comments-on-own-line.js:50]: Yea, I thought about this case and concluded in my head that we realistically won't encounter this situation.

Technically we could add a check if `node.lang === 'html'` (and make it a requirement that `html` codeblocks be marked with `html`) but this seems excessive 🤔

**Comment by BRKalow** [comment]
@thiskevinwang do you think this is something we'll really only be interested in applying to our TF docs?

**Comment by thiskevinwang** [comment]
> @thiskevinwang do you think this is something we'll really only be interested in applying to our TF docs?

@BRKalow I think it _wouldn't hurt_ to apply to all docs... but might be reasonable to apply only to TF docs

**Pull request closed without merging**